DRA scheduler: ResourceSlice update

pohly opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Making unschedulable pods schedulable again after ResourceSlice cluster events
was accidentally left out when adding structured parameters to Kubernetes 1.30.

#### Special notes for your reviewer:

All E2E tests were defined so that a driver starts first. A new test with a
different order (create pod first, wait for unschedulable, start driver)
triggered the bug and now passes.

While re-reviewing the event handler code, some drive-by changes were also made:
- reduce log verbosity in two cases
- fix feature gate check (irrelevant in practice, see commit message)

#### Does this PR introduce a user-facing change?
```release-note
DRA: when a DRA driver was started after creating pods which need resources from that driver, no additional attempt was made to schedule such unschedulable pods again. Only affected DRA with structured parameters.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4381
```

pohly committed:
DRA scheduler: fix feature gate check for PodSchedulingContext event

The event is only relevant when DRAControlPlaneController (= "classic DRA") is
enabled.

This change has no effect in practice because the only plugin using this event,
the dynamic resource plugin, also checks feature gates when asking for events
and correctly only asks for PodSchedulingContext events when
DRAControlPlaneController is enabled.

pohly committed:
DRA scheduler: reduce log verbosity

That a pod with no claims remains unschedulable on claim changes is a pretty
normal case. It should only be logged when debugging.

pohly committed:
DRA scheduler: fix re-scheduling after ResourceSlice changes

Making unschedulable pods schedulable again after ResourceSlice cluster events
was accidentally left out when adding structured parameters to Kubernetes 1.30.

All E2E tests were defined so that a driver starts first. A new test with a
different order (create pod first, wait for unschedulable, start driver)
triggered the bug and now passes.

aojea commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -491,6 +493,33 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
 	return framework.Queue, nil
 }
 
+// isSchedulableAfterClaimChange is invoked for add and update claim events reported by
```
nit s/isSchedulableAfterClaimChange/isSchedulableAfterResourceSliceChange/

pohly replied:
Fixed.

bart0sh commented:
/triage accepted
/priority important-soon
/cc @klueska

pohly commented:
/retest

pull-kubernetes-e2e-gce failed to come up. Other tests are passing now.

aojea commented:
not familiar enough with these codebase, it looks correct by diffing against the code around 
LGTM

pohly commented:
/assign @sanposhiho @alculquicondor 

@aojea did some preliminary review, but someone from SIG Scheduling has to have the final word and needs to approve.

alculquicondor commented:
/assign @macsko
for first pass

macsko commented:
/lgtm

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -445,7 +447,7 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
 		// foreachPodResourceClaim only returns errors for "not
 		// schedulable".
-		logger.V(4).Info("pod is not schedulable", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim), "reason", err.Error())
+		logger.V(6).Info("pod is not schedulable", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim), "reason", err.Error())
```
```suggestion
		logger.V(6).Info("Pod is not schedulable after claim change", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim), "reason", err.Error())
```

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -491,6 +493,33 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
+		// This is not an unexpected error: we know that
+		// foreachPodResourceClaim only returns errors for "not
+		// schedulable".
+		logger.V(6).Info("pod is not schedulable", "pod", klog.KObj(pod), "resourceSlice", klog.KObj(modifiedSlice), "reason", err.Error())
```
```suggestion
		logger.V(6).Info("Pod is not schedulable after resource slice change", "pod", klog.KObj(pod), "resourceSlice", klog.KObj(modifiedSlice), "reason", err.Error())
```

pohly replied:
Added, and also extended the same message in `isSchedulableAfterClaimChange`.

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -491,6 +493,33 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
 }
 
+// isSchedulableAfterResourceSliceChange is invoked for add and update slice events reported by
+// an informer. It checks whether that change made a previously unschedulable
```
Explain what exactly could make it schedulable according to this function.

It looks like we just care whether the pod claims exist, but we don't check any correlation between the claims and the ResourceSlice

pohly replied:
Updated.

We had this discussion on Slack (?) about the tradeoff between execution time for queuing hints vs. accuracy. Based on that discussion, I'm erring on the side of faster execution.

alculquicondor commented on test/e2e/dra/deploy.go:
```diff
@@ -77,69 +77,81 @@ type Nodes struct {
+// NewNodesNow is a variant of NewNodes which can be used inside a ginkgo.It.
+func NewNodesNow(ctx context.Context, f *framework.Framework, minNodes, maxNodes int) *Nodes {
+	nodes := &Nodes{}
+	nodes.beforeEach(ctx, f, minNodes, maxNodes)
```
```suggestion
	nodes.newNodes(ctx, f, minNodes, maxNodes)
```
?

Any other name that could summarize the logic in here?

pohly replied:
How about "init"?

pohly commented:
@alculquicondor: please take another look: https://github.com/kubernetes/kubernetes/compare/fb739de3cb27d5f8f5c21a9273b8be6d23c25bfb..e85d3babf0ff3c010dd8bae9329a2465c2c46527

pohly committed:
DRA scheduler: reduce verbosity of "Checked claim"

The original usage of `Allocate` was that the caller determines verbosity by
passing a logger with increased verbosity threshold into the call. Later that
was changed to using higher V levels inside `Allocated` itself, but the
"Checked claim" log call slipped through.

alculquicondor reviewed (commented):
/lgtm
/approve

pohly commented:
Let's also backport: https://github.com/kubernetes/kubernetes/pull/126873
